### ISSUE TITLE
Limited functionality of clone Space

### ISSUE THREAD

**Issue by DavidGoldwasser** [NREL/OpenStudio#4025]
Issue overview
--------------
Clone spaces works for for typical models, but I'm using a model with a number of additional objects that could be but are not handled by the clone.

Update: just wanted to clarify that my use case here is cloning from model_a to model_b, not within the same model.

## Current Behavior

- [x] OS:SurfaceProperty:ExposedFoundationPerimeter assigned to surfaces in the space are brought over with the clone but the surface name assignment is list
- [x] OS:SurfaceProperty:ConvectionCoefficients assigned to surfaces in the space are not brought over
- [x] OS:ShadingControl assigned to by sub-surfaces in the space are not brought over
## Expected Behavior

The space should be cloned over and be in a runnable state without loosing functionality. There may be support for additional objects beyond the ones listed
## Steps to Reproduce

I can provide OSM files (Apartment unit made from HPXML that are being merged into single OSM for URBANopt
https://github.com/urbanopt/urbanopt-example-geojson-project/blob/build-res-hpxml-v3-mf/measures/BuildResidentialModel/measure.rb

1.
2.
3.
4.

## Possible Solution
Review objects that have been added to space that were not original to OpenStudio when Clone method was written, and add support for them.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version):
 - Version of OpenStudio (if using an intermediate build, include SHA):

### Context
I'm having to write code to work around this to merge models together and maintain these objets in functional form. Could call this a bug or enhancement depending on viewpoint. It isn't critical for me for this release, but I think at some point soon an upgrade fo clone across object types to be in sync with new objects in OpenStudio would be great.

**Comment by DavidGoldwasser** [comment]
@kbenne thought you might be best to decide who should do this, if not you. For process I'm using to merge multiple OSM's into one OSM for multifamily URBANopt I will also be pushing clone for zones and air loops. I understand that for those clone isn't so start forward as it is hard to decide how deep to go and which objects belong to a terminal or a zone, in the case of something like a terminal. I expect I'll have to use some additional logic in my measure to match my specific use case.

I also need to find a way to maintain EMS code when merging multiple apartment units together.

I plan to work on HVAC and EMS functionality start on August 3rd

**Comment by DavidGoldwasser** [comment]
Just to add some details about `OS:SurfaceProperty:ExposedFoundationPerimeter`. While zone isn't optional for these they come in malformed without the zone field. My first work around was to check for nil zone and remove object
```
    # clean up objects that might be orphaned and warn user so they have option to correct
    model.getSurfacePropertyExposedFoundationPerimeters.each do |surf_prop|
      if surf_prop.surfaceName.nil? # isn't optional but field is ending up empty when comes over with spaces
        runner.registerWarning("surfacePropertyExposedFoundationPerimeters with handle of #{surf_prop.handle} isn't associated with a surface.")
        surf_prop.remove
      end
    end
```

I then improved this a bit to fix them in place, but had to rely on setString as there isn't an API to set the surface.
```
      hash[:surf_prop_exp_found_perim].each do |surf,surf_prop|
        target_surface = model.getSurfaceByName(surf.name.get).get
        new_surf_prop = surf_prop.clone(model).to_SurfacePropertyExposedFoundationPerimeter.get
        new_surf_prop.setString(1,target_surface.handle.to_s)
      end
```

**Comment by DavidGoldwasser** [comment]
Here is work around for convection coefficients. In both cases here an above when using a hash. The hash contains information from the source model of the cloned object that I use to help rebuild in the new model
```
      hash[:surf_prop_conv_coef].each do |surf,surf_prop|
        target_surface = model.getSurfaceByName(surf.name.get.to_s).get
        new_surf_prop = surf_prop.clone(model).to_SurfacePropertyConvectionCoefficients.get
        new_surf_prop.setSurface(target_surface)
      end 
```

**Comment by DavidGoldwasser** [comment]
One other item to add to list.

[ ] OS:Foundation:Kiva, (This is referred to by surfaces)
[ ] OS:Foundation:Kiva:Settings (related item, but this can only have single instance, so maybe it should only be cloned if the model the prior object is being cloned to doesn't already have the settings objet

**Comment by kbenne** [comment]
@DavidGoldwasser FoundationKivaSettings is a child of the Site and therefore cloned at that level. I guess I could check if there is a KivaSettings object when the Space is cloned and if not bring the KivaSettings over at that time. I kind of feel like this is a bit magical though and I would prefer to leave that alone. I have made a change to bring FoundationKiva (not settings) over with the `Surface::clone`

**Comment by DavidGoldwasser** [comment]
While I would welcome the approach that clones setting if it doesn't already exist, since it is child of site, I can understand wanting to leave it alone, so I'm ok with leaving KivaSettings alone.

**Comment by kbenne** [comment]
hmm... if there are not KivaSettings but FoundationKiva is used, does the simulation fail? That might be justification for cloning settings over with FoundationKiva

**Comment by DavidGoldwasser** [comment]
Yes, it will fail. I think there might be a way to do getKivaSettings similar to getSite that can make it on the fly if it doesn't already exist, but I cloned it in my measure so I can maintain the specific settings from the mdoel that used Kiva instead of making a new one with OpenStudio defaults.

**Comment by tijcolem** [comment]
closed via #4073
